Issue #4274: ImportControl: strategyOnMismatch property

timurt opened the pull request:
https://github.com/checkstyle/checkstyle/issues/4274

romani commented:
please rebase on latest code to resolve TC problem, TC problems are resolved in our master.

Other failures are caused by your changes, please make them green

MEZk commented:
@timurt 
CIs should be green.

Vladlis commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControl.java:
```diff
@@ -256,10 +284,31 @@ public AccessResult checkAccess(final String inPkg, final String forImport) {
         else if (parent == null) {
             // we are the top, so default to not allowed.
-            result = AccessResult.DISALLOWED;
+            if (strategyOnMismatch != null &&
```
I'd use `==` foe enum values

timurt replied:
@Vladlis @MEZk thanks, done

rnveach commented:
@timurt Please fix up your branch. There should be only 1 commit, your work.

timurt commented:
@rnveach Yeah, I know, thanks, trying to fix it right now

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControlOption.java:
```diff
@@ -0,0 +1,45 @@
+ * @author Timur Tibeyev
+ * @see ImportControlCheck
+ */
+public enum ImportControlOption {
```
name of class if too general.
it describe strategy of module on mismatch - MismatchStrategy.

original idea of property 'strategyOnMismatch' ....... looks like `mismatchStrategy` is more compact, if not disagreements - please update in whole code.

timurt replied:
renamed to MismatchStrategy

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControlOption.java:
```diff
@@ -0,0 +1,45 @@
+
+    /**
+     * In a case if matching allow/disallow rule was not found,
+     * check should go up to the parent and continue checking.
```
"check should go" ==> "decision should be delegated to"

This class should not care about where it is used (Check), it describes strategies.

timurt replied:
done

romani commented on src/main/resources/com/puppycrawl/tools/checkstyle/checks/imports/import_control_1_3.dtd:
```diff
@@ -0,0 +1,92 @@
+
+<!DOCTYPE module PUBLIC
+    "-//Puppy Crawl//DTD Import Control 1.2//EN"
+    "http://checkstyle.sourceforge.net/dtds/import_control_1_2.dtd">
```
should be 1_3.dtd

timurt replied:
done

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControlOption.java:
```diff
@@ -0,0 +1,45 @@
+     * In a case if matching allow/disallow rule was not found,
+     * check should go up to the parent and continue checking.
+     */
+    DELEGATE_TO_PARENT
```
as we keep this strategy as default, it should be first in Enum.
I think it do make sense, it will auto bind it to 0 value.

timurt replied:
done

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControl.java:
```diff
@@ -106,6 +123,17 @@
+     * @param subPkg the sub package name.
+     * @param regex flags interpretation of subPkg as regex pattern.
+     */
+    ImportControl(final ImportControl parent, final String subPkg, final boolean regex) {
```
@rnveach , what do you think to create a Check to forbid "final" on arguments ?
we have a guard Check that catch reusage of argument, so this finals are just make code verbose.

rnveach replied:
@romani There are valid uses of `final` on parameters or for-loop variables. If it is used in an anonymous class, it must be declared final or it can be referenced directly inside the class.
That said, we should have a check that produces a violation if it isn't used in an anonymous class.

romani replied:
`final` is also possible to place on catch arguments `catch(final Exception ex) { ... }`

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControlLoader.java:
```diff
@@ -227,6 +251,26 @@ private ImportControl getRoot() {
+        else if ("disallowed".equals(returnValue)) {
+            strategyOnMismatch = MismatchStrategy.DISALLOWED;
+        }
+        return strategyOnMismatch;
```
I do not like that you return NULL there, NULL is old way how engineer described uncertainty.
we always know strategy, there is default. NULL is not allowed

timurt replied:
I made this way because this method can be applied for both "import-control" and "subpackage" property. So if NULL is not allowed, I will have to make two separate methods for "import-control" and "subpackage", because they have different default strategies.

Also in the beginning of this class I have if condition

```
if (strategyOnMismatch == null) {
                 stack.push(new ImportControl(pkg, regex));
 }
 else {
                 stack.push(new ImportControl(pkg, regex, strategyOnMismatch));
}
```
and inside ImportControl.java I have constructor

```
ImportControl(final String pkgName, final boolean regex) {
        this(pkgName, regex, MismatchStrategy.DISALLOWED);
}
```
As you can see NULL object will not pass to the ImportControl object and default strategy will be initialized with proper value.
I can rewrite code, but I don't understand why my solution is not correct

timurt replied:
rewrote the code

romani commented on src/main/resources/com/puppycrawl/tools/checkstyle/checks/imports/import_control_1_3.dtd:
```diff
@@ -0,0 +1,92 @@
+  regular expression.
+
+  strategyOnMismatch - Strategy in a case if matching allow/disallow rule
+  was not found. Possible values: allowed, disallowed, delegateToParent
```
please explain user, if not defined it is delegateToParent.

timurt replied:
done

romani commented on src/xdocs/config_imports.xml:
```diff
@@ -983,6 +983,60 @@ public class InputIllegalImport { }
+          <code>strategyOnMismatch</code> property is attribute of <code>import-control</code> and <code>subpackage</code> tags.
+          Property can have following values:
+          <ul>
+            <li>delegateToParent (only for <code>subpackage</code>)</li>
```
it is hard to understand this "only for ...."  please clearly define what is allowed for import-control and subpackage .

timurt replied:
done

timurt commented:
@romani 

> 1. Please also suggest usage of new feature in our config file, as we have to eat food that we produce.

I made some changes inside config/import-control.xml.
Some improvements:
1. I removed all java.util.stream disallows and replaced them by allowing only java.util.stream.Collectors class.
2. Isolated "gui" and "doclets" packages with strategyOnMismatch="disallowed", because they do not need all import from parent.
3. Optimized rules by combining some of them together using regexps or by adding "exact-match" attribute.

Notes
1. I have noticed, that attribute "exact-match" was not used at all.
2. Also, some package rules has "local-only" attribute, which is useless, because they actually do not have subpackages, so attribute is redundant.
3. I have no clear understanding why some imports are allowed and some are disallowed. May be we should remove "local-only" restriction for `java.nio` and `java.lang.reflect` packages.

rnveach commented:
> I removed all java.util.stream disallows and replaced them by allowing only java.util.stream.Collectors class.

@timurt Did you read the comment above the disallows? Allowing those classes causes issues with cobertura. They CANNOT be allowed.
You need to verify the new class you are allowing doesn't cause issues too, otherwise it has to be disallowed too.

timurt commented:
@romani 
> Allowing those classes causes issues with cobertura. They CANNOT be allowed.

Yeah I have read issue, but import `java.util.stream.Collectors` is being used in many classes and it is only class from `java.util.stream` package is being used. So instead of having several disallows, there is only one allow. Should I revert it?

rnveach commented:
> So instead of having several disallows, there is only one allow. Should I revert it?

Yes.
It is specifically written so we don't forget. If users don't see the disallow specifically written, no one will remember and then they will just add them back in as allows.

romani commented on src/xdocs/config_imports.xml:
```diff
@@ -997,6 +997,64 @@ public class InputIllegalImport { }
         <p>
+          In the next examples usage of <code>strategyOnMismatch</code> property is shown.
+          This property defines strategy in a case when no matching allow/disallow rule was found.
+          <code>strategyOnMismatch</code> property is attribute of <code>import-control</code> and <code>subpackage</code> tags.
```
should be "Property <code>strategyOnMismatch</code> " . It is better to not start sentence from code block.

timurt replied:
done

romani commented on src/xdocs/config_imports.xml:
```diff
@@ -997,6 +997,64 @@ public class InputIllegalImport { }
+        </source>
+
+        <p>
+          When <code>strategyOnMismatch</code> has <b>allowed</b> or <b>disallowed</b> value for <code>subpackage</code> tag, it makes <code>subpackage</code> isolated from parent rules.
```
"<b>allowed</b> or <b>disallowed</b>" looks weird in html, please convert them to "code" as we use in all other cases "code" for values.

timurt replied:
done

jochenvdv commented:
@romani @timurt FYI:

In #3816 I decided to keep `local-only` for packages that don't have subpackages because they might get subpackages in the future and we want to be as specific as possible. Without `local-only` we would automatically allow future imports from those new subpackages which we might not want.

rnveach commented on config/import-control.xml:
```diff
@@ -146,13 +146,15 @@
-    <allow class="com.puppycrawl.tools.checkstyle.TreeWalker" local-only="true"/>
+    <allow pkg="java.util" exact-match="true"/>
+    <allow pkg="com\.puppycrawl\.tools\.checkstyle\.(api|utils)" regex="true"/>
+    <allow class="com\.puppycrawl\.tools\.checkstyle\.(TreeWalker|JavadocDetailNodeParser)" regex="true"/>
```
> 3. Please do not optimize our config by regexp. No need to keep it compact. A little of verbosity is good there.

I thought we were removing regular expressions?

MEZk replied:
There is no need in regular expression. @romani is right. Lets keep 'allow' elements for each import.

romani replied:
ok, lets keep non regexp form.

timurt replied:
done

rnveach commented on src/xdocs/config_imports.xml:
```diff
@@ -997,6 +997,64 @@ public class InputIllegalImport { }
+          Property can have following values for <code>import-control</code> tag:
+          <ul>
+            <li>disallowed (default value) - if there is no matching allow/disallow rule in any of the subpackages, including the root level (import-control), then the import is allowed.</li>
+            <li>allowed - if there is no matching allow/disallow rule in any of the subpackages, including the root level, then the import is disallowed.</li>
```
Don't you have the ending swapped for each value?

> disallowed ... then the import is allowed.
> allowed ... then the import is disallowed

Shouldn't `disallowed` say "then the import is disallowed"?
The next section doesn't have this issue.

timurt replied:
done

MEZk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControlLoader.java:
```diff
@@ -227,6 +247,37 @@ private ImportControl getRoot() {
+    private static MismatchStrategy getStrategyForSubpackage(final Attributes attributes) {
+        final String returnValue = attributes.getValue(STRATEGY_ON_MISMATCH_ATTRIBUTE_NAME);
+        MismatchStrategy strategyOnMismatch = MismatchStrategy.DELEGATE_TO_PARENT;
+        if (STRATEGY_ON_MISMATCH_ALLOWED_VALUE.equals(returnValue)) {
```
@rnveach @romani 
Can we use MismatchStrategy enum here instead of the sting (STRATEGY_ON_MISMATCH_ALLOWED_VALUE) ?

romani replied:
String type should stay in this code , but we could initialize STRATEGY_ON_MISMATCH_ALLOWED_VALUE from enums `toString()`.
@timurt , please do.

romani replied:
No we cannot do this as delegateToParent is not in the same spelling as DELEGATE_TO_PARENT

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControlLoader.java:
```diff
@@ -66,12 +70,25 @@
+
+    /** Value "disallowed" for attribute 'strategyOnMismatch'. */
+    private static final String STRATEGY_ON_MISMATCH_DISALLOWED_VALUE = "disallowed";
+
```
there should be constant for "delegateToParent" , default value is still valid value.

timurt replied:
we do not have to use this value inside class since default value initialized using Enum

romani replied:
confirmed , this is details of implementation of method `private static MismatchStrategy getStrategyForSubpackage(final Attributes attributes) `

romani commented:
@timurt , there will be release at this Sunday - please find time to do this minor changes.

rnveach commented on config/import-control.xml:
```diff
@@ -146,13 +146,17 @@
     <allow class="com.puppycrawl.tools.checkstyle.TreeWalker" local-only="true"/>
     <allow class="com.puppycrawl.tools.checkstyle.gui.MainFrameModel.ParseMode"
-          local-only="true"/>
+           local-only="true"/>
```
undo superfluous change.

romani replied:
it is attempt to align name of attributes, it is ok.
